### PR TITLE
Fake data improvements

### PR DESCRIPTION
**Pull request by MauriceReeves-usds**
This PR changes the way we do fake data generation including allowing us to use the mappers in our generation of fake data.

## Changes
- Adds fake data mapper usage
- Tests for the above
- Small tweaks for the FL schema that resulted from testing data generation

## Checklist
- [x] Tested locally?
- [ ] Added new dependencies?
- [ ] Updated the docs?
- [x] Added a test?
- [x] Did you check for sensitive data, and remove any?
- [ ] Are additional approvals needed for this change?
- [ ] Are there potential vulnerabilities or licensing issues with any new dependencies introduced?

## Fixes 

- #issue

## To Be Done

- #issue

### CODE REVIEW

**Review by jimduff-usds — COMMENTED** [review]
FYI, I'm seeing an exception with a run of "quick-test.sh":
```
Now send that fake FL data through the router.
Exception in thread "main" java.lang.IllegalStateException: Schema Error: Could not find element 'patient_age' for mapper 'concat' from 'patient_age_and_units'.
	at gov.cdc.prime.router.serializers.CsvSerializer$mapRow$3.invoke(CsvSerializer.kt:219)
	at gov.cdc.prime.router.serializers.CsvSerializer.mapRow(CsvSerializer.kt:241)
	at gov.cdc.prime.router.serializers.CsvSerializer.read(CsvSerializer.kt:71)
	at gov.cdc.prime.router.serializers.CsvSerializer.read$default(CsvSerializer.kt:52)
	at gov.cdc.prime.router.serializers.CsvSerializer.read(CsvSerializer.kt:44)
	at gov.cdc.prime.router.cli.RouterCli$run$inputReport$1.invoke(main.kt:207)
	at gov.cdc.prime.router.cli.RouterCli$run$inputReport$1.invoke(main.kt:35)
	at gov.cdc.prime.router.cli.RouterCli.readReportFromFile(main.kt:80)
	at gov.cdc.prime.router.cli.RouterCli.run(main.kt:206)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:171)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:16)
	at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:395)
	at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:392)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:410)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:435)
	at gov.cdc.prime.router.cli.MainKt.main(main.kt:270)
*** ERROR ***: ./prime did not generate a file that matches /fl
```

**Comment by RickHawesUSDS** [comment]
No exceptions for me after Chris's changes running the all the quick tests. Generated a STRAC schema fake file without problems.

File: prime-router/metadata/schemas/FL/fl-covid-19.schema
Comment (RickHawesUSDS) [inline prime-router/metadata/schemas/FL/fl-covid-19.schema:264]: Does `delimiter` only apply to csv fields or the concat mapper? If so, I would like it to be put into that. We can talk about it.

File: prime-router/src/main/kotlin/cli/main.kt
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/cli/main.kt:10]: There is a Kotlin style guide violation against using wildcards.

File: prime-router/src/test/kotlin/SimpleReportTests.kt
Comment (RickHawesUSDS) [inline prime-router/src/test/kotlin/SimpleReportTests.kt:137]: Is this a mistake. If you want to disable a test, please use @Ignore. But why disable this test?

**Review by RickHawesUSDS — CHANGES_REQUESTED** [review]
Lot's of good improvements in this PR. Just a couple of small details that I would like fixed.

File: prime-router/metadata/schemas/FL/fl-covid-19.schema
Comment (MauriceReeves-usds) [reply prime-router/metadata/schemas/FL/fl-covid-19.schema:264]: It applies to the concat mapper. I like having it in Element because we inject Element into the mapper, and I feel like the Element should carry the information about how it wants to be presented/constructed when it's injected.

File: prime-router/src/main/kotlin/cli/main.kt
Comment (MauriceReeves-usds) [reply prime-router/src/main/kotlin/cli/main.kt:10]: I will fix this.

File: prime-router/src/test/kotlin/SimpleReportTests.kt
Comment (MauriceReeves-usds) [reply prime-router/src/test/kotlin/SimpleReportTests.kt:137]: The test was failing during my changes. I need to turn this back on and check it.

**Comment by MauriceReeves-usds** [comment]
@jimduff-usds @RickHawesUSDS - I fixed the remaining issues. If I can get an approval, we can merge this in.

File: prime-router/src/test/kotlin/SimpleReportTests.kt
Comment (jimduff-usds) [reply prime-router/src/test/kotlin/SimpleReportTests.kt:141]: After that fix, two tests still fail.   The pima test, and the FL test.      I suspect the FL test is failing for the same reason as the ./quick-test.sh FL test has been failing.    Pima looks like another tweak needed, but I didn't look into it.

**Review by jimduff-usds — CHANGES_REQUESTED** [review]
@mauricereeves , could you do another pass at making those SimpleReport tests work?
At least one of the changes was easy (changing "Id" to "ID"), and one looks to me like an issue that we found a week or so ago, maybe an issue with "format" in the FL data.

I'm happy to assist, if you want to set up some time - with the caveat I'll be out on Friday.

File: prime-router/src/test/kotlin/SimpleReportTests.kt
Comment (MauriceReeves-usds) [reply prime-router/src/test/kotlin/SimpleReportTests.kt:141]: Yup. I'll look at this now.

**Comment by MauriceReeves-usds** [comment]
@jimduff-usds @RickHawesUSDS the SimpleReportTests are passing now and so is `quick-test.sh`. Can you take another look. There is a bug I've filed that's related to this as well: #323

**Comment by MauriceReeves-usds** [comment]
Once the PR is approved we can close this out, but there is a bug that was uncovered (#323) and we might want to wait to merge until we've confirmed the bug and closed 323 as well.

**Comment by MauriceReeves-usds** [comment]
Updated tag to specify blocked by #323

**Comment by MauriceReeves-usds** [comment]
@jimduff-usds @RickHawesUSDS I updated the value set logic so it allows for replacing a value in place, and will map correctly either way so if it receives the replacement code OR the replaced code it will not fail. I restored the alt values into the FL schema and rerun `quick-test.sh` and our test suite and there's no failures. The generated fake data looks correct as well. I'd like to merge this in officially. I'd still like us to look at #323 but it's not blocking this PR now.

File: prime-router/src/main/kotlin/Element.kt
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/Element.kt:265]: this when(format) else when(format) is very awkward. Can we refactor?

File: prime-router/src/main/kotlin/Element.kt
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/Element.kt:534]: See my when commenting above.

File: prime-router/src/main/kotlin/ValueSet.kt
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/ValueSet.kt:47]: AltValue is kind of a replaces of the main value set. When do we need a replaces vs an altValue? Isn't this a map?

**Review by RickHawesUSDS — CHANGES_REQUESTED** [review]

File: prime-router/src/main/kotlin/ValueSet.kt
Comment (MauriceReeves-usds) [reply prime-router/src/main/kotlin/ValueSet.kt:47]: It would probably be easier to have this discussion in a meeting, but the idea here is for something like the Florida file where FL expects `Y, N, or U` instead of the `Y, N, UNK`. By specifying that `U` is a drop in for the `UNK` we can correctly translate values when they come in, either a `U` or a `UNK`

File: prime-router/src/main/kotlin/Element.kt
Comment (MauriceReeves-usds) [reply prime-router/src/main/kotlin/Element.kt:265]: Yup, I will take a look.

File: prime-router/src/main/kotlin/Element.kt
Comment (MauriceReeves-usds) [reply prime-router/src/main/kotlin/Element.kt:534]: I will take a look.

File: prime-router/src/main/kotlin/ValueSet.kt
Comment (RickHawesUSDS) [reply prime-router/src/main/kotlin/ValueSet.kt:47]: So, just to document our conversation. We agree to accept `replace` for now, but to generalize altValues to work for mapping display and code values in the future.

**Review by RickHawesUSDS — APPROVED** [review]
Thanks for the improvements. LGTM.

**Review by jimduff-usds — APPROVED** [review]
Just re-an quick-test in the Fake_Data_Improvements branch - not seeing any exceptions now.

Also tested generating PDI data in FL, then routing that to FL output schema:
```
./prime data --target-state FL --input-schema primedatainput/pdi-covid-19 --input-fake 100 --output-dir junk/
./prime data --route --output-dir ./junk --input ./junk/pdi-covid-19-7a805b97-b607-42ae-8551-a5e6d1758099-20210125124719.csv --route --input-schema primedatainput/pdi-covid-19 
```

which also worked.

**Pull request merged**
